[addons] remove cpluff

AlwinEsch opened the pull request:
## Description
This request removes the use of "cpluff" in Kodi.

**Why should be cpluff removed:**
cpluff was designed to run a complete addon management. This includes not only loading the addon data, but also using and running these addons.

Kodi has only used the loading of addon.xml. This can also be done quite easily by yourself.

Furthermore, the performance of cpluff is rather bad and Kodi must always retry the data stored in it to create its matching data structure. Therefore, the data passes through twice in the system, once in cpluff and once in Kodi.

Also, cpluff does not allow to easily work together multiple addon types in an addon itself, see Binary addons. There is only information related to the respective addon type and no other type in it.

**What does this:**
It searches the addons stored in the system or reads the data given by the repository and then creates it from the found xml file. Or by using [CAddonInfoBuilder::CFromDB](https://github.com/xbmc/xbmc/pull/14908/files#diff-d468e80a368d09a27acf00bab3edf35fR29) from database.

For local addons, the information created with [CAddonInfoBuilder::Generate](https://github.com/xbmc/xbmc/pull/14908/files#diff-11ac4eb50187a8acb45199c341852086R51) is then stored in the manager (instead of in cpluff). In the case of a request, the associated CAddonInfo will then be issued.

In the event of a change, this list will be replaced or an entry removed (as before). Because these are given as shared memory, conflicts are ruled out if used elsewhere (similar to the previous ones where the classes were copied instead).

**What changed here:**
This contains only the minimal changes and the access to the AddonManager remains as before.

Only the "FromExtension" is removed at the respective addon classes and replaced with a new path.

Furthermore, the CAddonInfo is no longer passed with a static address but with a shared pointer because the CAddonInfo is already stored in the AddonManager and must remain there.

This is another try. But definitely something has to be done because with cpluff, Kodi comes to its limits and the previous code (also by me for advanced abilities) is very disgusted.

Please do not come back with "no, no", if it does not fit, help me or bring a good one yourself.

## Motivation and Context
To finally make the addon system more extensible and secure

## How Has This Been Tested?
- [X] Test code system added
- [x] Test with mostly all available addons
- [x] Test automatic updates
- [x] Test uninstall / update to becomes not more bad
- [x] Test with bad addon.xml's
- [x] Test on all OS's (related e.g. to binary download on Windows)
- [x] Confirm current binary addon system works normal
- [X] Repository addon install over ZIP Ok (still need Kodi restart like before, to download included addons)

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [X] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch committed:
[addons] change from CAddonInfo from static to a shared pointer

This is necessary because when removing the cpluff the CAddonInfo is taken as replaced and is stored in the CAddonManager as a shared pointer in a list.

AlwinEsch committed:
[addons] add to CAddon construction the types as value

This is necessary because after removing cpluff the CAddonInfo contains the data of all types available in the addon.

AlwinEsch commented:
Also would like to ask if it is possible to add an agile branch here in the xbmc repo, as additional changes from my side are necessary (for example, to reorganize the PVR and game or to improve the addon system)?

This can then contain this and other changes.

AlwinEsch commented:
Found one bug in old system, by installing skin.grid, the dependencies of a dependency becomes not installed.

AlwinEsch commented:
Got a mistake and can not figure out why.

When loading the "skin.pellucid" which uses the "script.skinshortcuts" to create a "script-skinshortcuts-includes.xml" (even with other skins that use shourtcuts), the created file will not be found, even though they are was created and exists.

[CGUIIncludes :: Load_Internal](https://github.com/xbmc/xbmc/blob/master/xbmc/guilib/GUIIncludes.cpp#L113) then fails loading this file and the skin is not usable.

After a restart of Kodi it works.

The old way about cpluff went, but can not explain why. Could possibly imagine that something was done in it, that Kodi's file management was updated.

Are there any known errors that sometimes a skin with "script.skinshortcuts" usage does not work directly? And does anyone have an idea what that could be a mistake?

Here the Log: https://paste.kodi.tv/wacuxaripe.kodi

<details><summary>The log entries who confuse me:</summary>
<p>

```
08:20:36.182 T:139902097560000   ERROR: Control 9000 in window 10000 has been asked to focus, but it can't
08:20:36.184 T:139900842145536   DEBUG: -->Python Interpreter Initialized<--
08:20:36.184 T:139900842145536   DEBUG: CPythonInvoker(1, special://home/addons/script.skinshortcuts/default.py): the source file to load is "special://home/addons/script.skinshortcuts/default.py" ("/home/alwin/.kodi/addons/script.skinshortcuts/default.py")
08:20:36.186 T:139900842145536   DEBUG: CPythonInvoker(1, special://home/addons/script.skinshortcuts/default.py): setting the Python path to /home/alwin/.kodi/addons/script.skinshortcuts:/home/alwin/.kodi/addons/script.module.simplejson/lib:/home/alwin/.kodi/addons/script.module.unidecode/lib:/usr/lib/python2.7:/usr/lib/python2.7/plat-x86_64-linux-gnu:/usr/lib/python2.7/lib-tk:/usr/lib/python2.7/lib-old:/usr/lib/python2.7/lib-dynload:/usr/local/lib/python2.7/dist-packages:/usr/lib/python2.7/dist-packages:/usr/lib/python2.7/dist-packages/gtk-2.0:/usr/lib/python2.7/dist-packages/wx-3.0-gtk3
08:20:36.186 T:139900842145536   DEBUG: CPythonInvoker(1, special://home/addons/script.skinshortcuts/default.py): entering source directory /home/alwin/.kodi/addons/script.skinshortcuts
08:20:36.186 T:139900842145536   DEBUG: CPythonInvoker(1, special://home/addons/script.skinshortcuts/default.py): instantiating addon using automatically obtained id of "script.skinshortcuts" dependent on version 2.20.0 of the xbmc.python api
08:20:36.277 T:139902097560000 WARNING: CGLContextEGL::SwapBuffers: last msc time greater than interval
...
HERE ------------> 08:20:36.360 T:139900842145536    INFO: CPythonInvoker(1, special://home/addons/script.skinshortcuts/default.py): script successfully run
HERE ------------> 08:20:36.360 T:139900842145536   DEBUG: onExecutionDone(1, special://home/addons/script.skinshortcuts/default.py)
08:20:36.363 T:139900842145536    INFO: Python interpreter stopped
08:20:36.363 T:139900842145536   DEBUG: Thread LanguageInvoker 139900842145536 terminating
08:20:36.377 T:139902097560000 WARNING: CGLContextEGL::SwapBuffers: last msc time greater than interval (1)
08:20:36.379 T:139902097560000    INFO: Unloading old skin ...
08:20:36.380 T:139902097560000   DEBUG: ------ Window Deinit () ------
08:20:36.380 T:139902097560000   DEBUG: ------ Window Deinit (DialogExtendedProgressBar.xml) ------
08:20:36.380 T:139902097560000   DEBUG: ------ Window Deinit (Home.xml) ------
08:20:36.384 T:139902097560000  NOTICE:   load skin from: special://home/addons/skin.pellucid/ (version: 2.0.2)
08:20:36.384 T:139900808574720   DEBUG: Thread Timer 139900808574720 terminating
08:20:36.385 T:139902097560000    INFO:   load fonts for skin...
08:20:36.386 T:139902097560000    INFO: Loading skin includes from /home/alwin/.kodi/addons/skin.pellucid/1080i/Includes.xml
HERE ------------> 08:20:36.393 T:139902097560000    INFO: Error loading include file special://home/addons/skin.pellucid/1080i/script-skinshortcuts-includes.xml: Failed to open file (row: 0, col: 0)
08:20:36.394 T:139902097560000    INFO: Loading fonts from special://home/addons/skin.pellucid/1080i/Font.xml
08:20:36.402 T:139902097560000   DEBUG: LocalizeStrings: loaded 244 strings from file /home/alwin/.kodi/addons/skin.pellucid/language/English/strings.po
08:20:36.402 T:139902097560000    INFO:   load new skin...
08:20:36.402 T:139902097560000    INFO: Loading custom window XMLs from skin path special://home/addons/skin.pellucid/1080i
08:20:36.405 T:139902097560000   DEBUG: Load Skin XML: 2.85ms
08:20:36.405 T:139902097560000    INFO:   initialize new skin...
```

</p>
</details>

AlwinEsch reviewed (commented):
-

AlwinEsch commented on xbmc/filesystem/File.cpp:
```diff
@@ -278,7 +278,12 @@ bool CFile::Open(const CURL& file, const unsigned int flags)
+        if (!XFILE::CFile::Exists(url2, false))
+          return false;
+        g_directoryCache.AddFile(url2.Get());
+      }
```
This have fixed my skin load problem with the "[vfs]" commit and works, but not sure if this way is OK?

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
Now also tested under windows and runs there perfectly.

For the review, go through the individual commits. Have these subdivided that they could almost be seen as a separate request.

By removing the cpluff it looks huge, but if you ignore the "[addons] remove cpluff complete" commit you will see the needed changes.

hudokkow commented:
Hey Alwin.

I took this for a spin today. Built it along with all our bin addons, and couldn't see any problems. Tested vdr and tvheadend addons with my backends and I must say, great work!

Everything I tested works flawlessly - as far as I can see - and without hiccups.

Kudos! :+1:

AlwinEsch commented:
Hi @hudokkow, Thanks a lot for testing! 

The code here must be tackled soon so that he might be able to get started by Kodi M ...

stefansaraev commented:
it is time ;)

AlwinEsch commented:
Can be this checked and reviewed to know it can come in.

notspiff reviewed (approved):
Nothing spotted on a fairly rigorous review. Let's just deal with fallout.

notspiff commented:
jenkins build this please

hudokkow commented:
As far as I can see builds are failing because they introduce new warnings.

Rechi reviewed (commented):
Compiling specific commits is broken for `[addons] add primitives to create CAddonInfo without cpluff`, `[addons] change from CAddonInfo from static to a shared pointer`, `[addons] add to CAddon construction the types as value`, `[addons] move on without "cpluff"`, `[addons] remove cpluff complete`, `[addons] adds support for various description languages`.
Can you fix that?

AlwinEsch commented:
Thanks a lot @Rechi @notspiff !

Have updated related to your comments and hope have nothing overseen.

Also does the build for every commit work now.

AlwinEsch committed:
[addons] move on without "cpluff"

This converts the system to use the internal addons build system.

The cpluff is removed in the following commit.

notspiff commented:
jenkins build this please

ksooo commented:
May I suggest to include this PR in @MilhouseVH builds for some time before we're going to merge this?

AlwinEsch commented:
@ksooo How you mean @MilhouseVH builds? Something on his repo?

ksooo commented:
@MilhouseVH daily releases LibreELEC Matrix test builds and is so kind to include not-yet-merged Kodi PRs for testing purposes if we ask him to do so. And that is what I just did. Milhouse's builds have been proven as a very good source for early feedback.

MilhouseVH commented:
I'm happy to include this PR for testing in nightly builds, however it doesn't currently apply on top of latest Kodi master and may require a rebase:
```
neil@nm-linux:~/projects/xx.git$ gs
On branch master
Your branch is up-to-date with 'origin/master'.

nothing to commit, working tree clean
neil@nm-linux:~/projects/xx.git$ git pull
Already up-to-date.
neil@nm-linux:~/projects/xx.git$ wget -q https://github.com/xbmc/xbmc/pull/14908.patch -O-|git apply -v
Checking patch cmake/treedata/common/subdirs.txt...
Checking patch xbmc/addons/CMakeLists.txt...
Checking patch xbmc/addons/IAddon.h...
Checking patch xbmc/addons/AddonInfo.cpp => xbmc/addons/addoninfo/AddonInfo.cpp...
Checking patch xbmc/addons/AddonInfo.h => xbmc/addons/addoninfo/AddonInfo.h...
Checking patch xbmc/addons/addoninfo/CMakeLists.txt...
Checking patch xbmc/addons/binary-addons/BinaryAddonBase.h...
Checking patch xbmc/addons/binary-addons/BinaryAddonType.h...
Checking patch xbmc/utils/FileExtensionProvider.h...
Checking patch xbmc/addons/addoninfo/AddonExtensions.cpp...
Checking patch xbmc/addons/addoninfo/AddonExtensions.h...
Checking patch xbmc/addons/addoninfo/AddonInfo.cpp...
Checking patch xbmc/addons/addoninfo/AddonInfo.h...
Checking patch xbmc/addons/addoninfo/AddonInfoBuilder.cpp...
Checking patch xbmc/addons/addoninfo/AddonInfoBuilder.h...
Checking patch xbmc/addons/addoninfo/AddonType.cpp...
Checking patch xbmc/addons/addoninfo/AddonType.h...
Checking patch xbmc/addons/addoninfo/CMakeLists.txt...
Checking patch xbmc/addons/binary-addons/BinaryAddonExtensions.cpp...
Checking patch xbmc/addons/binary-addons/BinaryAddonExtensions.h...
Checking patch xbmc/addons/Addon.cpp...
Checking patch xbmc/addons/Addon.h...
Checking patch xbmc/addons/AddonBuilder.cpp...
Checking patch xbmc/addons/AddonBuilder.h...
Checking patch xbmc/addons/AddonManager.h...
Checking patch xbmc/addons/ContextMenuAddon.cpp...
Checking patch xbmc/addons/ContextMenuAddon.h...
Checking patch xbmc/addons/FontResource.cpp...
Checking patch xbmc/addons/FontResource.h...
Checking patch xbmc/addons/GameResource.cpp...
Checking patch xbmc/addons/GameResource.h...
Checking patch xbmc/addons/ImageResource.cpp...
Checking patch xbmc/addons/ImageResource.h...
Checking patch xbmc/addons/LanguageResource.cpp...
Checking patch xbmc/addons/LanguageResource.h...
Checking patch xbmc/addons/PVRClient.cpp...
Checking patch xbmc/addons/PVRClient.h...
Checking patch xbmc/addons/PluginSource.cpp...
Checking patch xbmc/addons/PluginSource.h...
Checking patch xbmc/addons/Repository.cpp...
Checking patch xbmc/addons/Repository.h...
Checking patch xbmc/addons/Resource.h...
Checking patch xbmc/addons/Scraper.cpp...
Checking patch xbmc/addons/Scraper.h...
Checking patch xbmc/addons/Service.cpp...
Checking patch xbmc/addons/Service.h...
Checking patch xbmc/addons/Skin.cpp...
Checking patch xbmc/addons/Skin.h...
Checking patch xbmc/addons/UISoundsResource.h...
Checking patch xbmc/addons/Webinterface.cpp...
Checking patch xbmc/addons/Webinterface.h...
Checking patch xbmc/addons/binary-addons/AddonDll.cpp...
Checking patch xbmc/addons/binary-addons/AddonDll.h...
Checking patch xbmc/addons/binary-addons/BinaryAddonBase.cpp...
Checking patch xbmc/addons/binary-addons/BinaryAddonBase.h...
Checking patch xbmc/addons/binary-addons/BinaryAddonManager.cpp...
Checking patch xbmc/addons/interfaces/GUI/AddonCallbacksGUI.cpp...
Checking patch xbmc/addons/interfaces/GUI/Window.cpp...
Checking patch xbmc/games/addons/GameClient.cpp...
Checking patch xbmc/games/addons/GameClient.h...
Checking patch xbmc/games/controllers/Controller.cpp...
Checking patch xbmc/games/controllers/Controller.h...
Checking patch xbmc/interfaces/legacy/WindowXML.cpp...
Checking patch xbmc/addons/Addon.cpp...
Checking patch xbmc/addons/Addon.h...
Checking patch xbmc/addons/AddonBuilder.cpp...
Checking patch xbmc/addons/ContextMenuAddon.cpp...
Checking patch xbmc/addons/ContextMenuAddon.h...
Checking patch xbmc/addons/FontResource.h...
Checking patch xbmc/addons/GameResource.cpp...
Checking patch xbmc/addons/ImageResource.cpp...
Checking patch xbmc/addons/ImageResource.h...
Checking patch xbmc/addons/LanguageResource.cpp...
Checking patch xbmc/addons/LanguageResource.h...
Checking patch xbmc/addons/PVRClient.cpp...
Checking patch xbmc/addons/PluginSource.cpp...
Checking patch xbmc/addons/PluginSource.h...
Checking patch xbmc/addons/Repository.cpp...
Checking patch xbmc/addons/Repository.h...
Checking patch xbmc/addons/Resource.h...
Checking patch xbmc/addons/Scraper.cpp...
Checking patch xbmc/addons/Scraper.h...
Checking patch xbmc/addons/Service.h...
Checking patch xbmc/addons/Skin.cpp...
Checking patch xbmc/addons/UISoundsResource.h...
Checking patch xbmc/addons/Webinterface.cpp...
Checking patch xbmc/addons/Webinterface.h...
Checking patch xbmc/addons/binary-addons/AddonDll.cpp...
Checking patch xbmc/addons/binary-addons/AddonDll.h...
Checking patch xbmc/games/addons/GameClient.cpp...
Checking patch xbmc/games/controllers/Controller.cpp...
Checking patch xbmc/addons/Addon.cpp...
Checking patch xbmc/addons/Addon.h...
Checking patch xbmc/addons/AddonBuilder.cpp...
Checking patch xbmc/addons/AddonBuilder.h...
Checking patch xbmc/addons/AddonDatabase.cpp...
Checking patch xbmc/addons/AddonDatabase.h...
Checking patch xbmc/addons/AddonManager.cpp...
Hunk #3 succeeded at 563 (offset 10 lines).
Hunk #4 succeeded at 577 (offset 10 lines).
Hunk #5 succeeded at 593 (offset 10 lines).
Hunk #6 succeeded at 624 (offset 10 lines).
Hunk #7 succeeded at 640 (offset 10 lines).
Hunk #8 succeeded at 652 (offset 10 lines).
Hunk #9 succeeded at 700 (offset 10 lines).
Hunk #10 succeeded at 795 (offset 10 lines).
Hunk #11 succeeded at 1148 (offset 10 lines).
Hunk #12 succeeded at 1160 (offset 10 lines).
error: while searching for:
    CLog::Log(cp_to_clog(level), "ADDON: cpluff: '%s' reports '%s'", apid, msg);
}

} /* namespace ADDON */


error: patch failed: xbmc/addons/AddonManager.cpp:1350
error: xbmc/addons/AddonManager.cpp: patch does not apply
Checking patch xbmc/addons/AddonManager.h...
Checking patch xbmc/addons/ContextMenuAddon.cpp...
Checking patch xbmc/addons/ContextMenuAddon.h...
Checking patch xbmc/addons/ImageResource.cpp...
Checking patch xbmc/addons/ImageResource.h...
Checking patch xbmc/addons/LanguageResource.cpp...
Checking patch xbmc/addons/LanguageResource.h...
Checking patch xbmc/addons/PluginSource.cpp...
Checking patch xbmc/addons/Repository.cpp...
Checking patch xbmc/addons/Repository.h...
Checking patch xbmc/addons/Scraper.cpp...
Checking patch xbmc/addons/Skin.cpp...
Checking patch xbmc/addons/Skin.h...
Checking patch xbmc/addons/Webinterface.cpp...
Checking patch xbmc/addons/Webinterface.h...
Checking patch xbmc/games/addons/GameClient.cpp...
Checking patch .gitignore...
Checking patch CMakeLists.txt...
Checking patch cmake/modules/FindCpluff.cmake...
Checking patch lib/cpluff/.gitignore...
Checking patch lib/cpluff/CMakeLists.txt...
Checking patch lib/cpluff/COPYRIGHT.txt...
Checking patch lib/cpluff/ChangeLog.txt...
Checking patch lib/cpluff/INSTALL.txt...
Checking patch lib/cpluff/Issues.txt...
Checking patch lib/cpluff/Makefile.am...
Checking patch lib/cpluff/README.txt...
Checking patch lib/cpluff/Roadmap.txt...
Checking patch lib/cpluff/autogen.sh...
Checking patch lib/cpluff/configure.ac...
Checking patch lib/cpluff/console/Makefile.am...
Checking patch lib/cpluff/console/cmdinput_basic.c...
Checking patch lib/cpluff/console/cmdinput_readline.c...
Checking patch lib/cpluff/console/console.c...
Checking patch lib/cpluff/console/console.h...
Checking patch lib/cpluff/doc/.htaccess...
Checking patch lib/cpluff/doc/Makefile.am...
Checking patch lib/cpluff/doc/common_style.css...
Checking patch lib/cpluff/doc/copyright.en.html...
Checking patch lib/cpluff/doc/cpluff_style.css...
Checking patch lib/cpluff/doc/download.en.html...
Checking patch lib/cpluff/doc/img/Makefile.am...
Checking patch lib/cpluff/doc/img/architecture.png...
Checking patch lib/cpluff/doc/img/architecture.svg...
Checking patch lib/cpluff/doc/img/pageborder.png...
Checking patch lib/cpluff/doc/index.en.html...
Checking patch lib/cpluff/doc/listconfirm/alreadyon.en.php...
Checking patch lib/cpluff/doc/listconfirm/confirm.en.php...
Checking patch lib/cpluff/doc/listconfirm/invalid.en.php...
Checking patch lib/cpluff/doc/listconfirm/noton.en.php...
Checking patch lib/cpluff/doc/listconfirm/subscribed-announcement.en.php...
Checking patch lib/cpluff/doc/listconfirm/unsubscribed-announcement.en.php...
Checking patch lib/cpluff/doc/lists.en.html...
Checking patch lib/cpluff/doc/repository.en.html...
Checking patch lib/cpluff/doc/robots.txt...
Checking patch lib/cpluff/docsrc/Makefile.am...
Checking patch lib/cpluff/docsrc/doxygen.css...
Checking patch lib/cpluff/docsrc/doxygen.footer...
Checking patch lib/cpluff/examples/COPYRIGHT.txt...
Checking patch lib/cpluff/examples/INSTALL.txt...
Checking patch lib/cpluff/examples/Makefile.am...
Checking patch lib/cpluff/examples/Makefile.nmake...
Checking patch lib/cpluff/examples/autogen.sh...
Checking patch lib/cpluff/examples/common.nmake...
Checking patch lib/cpluff/examples/configure.ac...
Checking patch lib/cpluff/examples/cpfile/Makefile.am...
Checking patch lib/cpluff/examples/cpfile/Makefile.nmake...
Checking patch lib/cpluff/examples/cpfile/README.txt...
Checking patch lib/cpluff/examples/cpfile/cpfile.bat...
Checking patch lib/cpluff/examples/cpfile/cpfile.in...
Checking patch lib/cpluff/examples/cpfile/plugins/Makefile.am...
Checking patch lib/cpluff/examples/cpfile/plugins/Makefile.nmake...
Checking patch lib/cpluff/examples/cpfile/plugins/cext/Makefile.am...
Checking patch lib/cpluff/examples/cpfile/plugins/cext/Makefile.nmake...
Checking patch lib/cpluff/examples/cpfile/plugins/cext/plugin.xml...
Checking patch lib/cpluff/examples/cpfile/plugins/core/Makefile.am...
Checking patch lib/cpluff/examples/cpfile/plugins/core/Makefile.nmake...
Checking patch lib/cpluff/examples/cpfile/plugins/core/classifiers.xsd...
Checking patch lib/cpluff/examples/cpfile/plugins/core/core.c...
Checking patch lib/cpluff/examples/cpfile/plugins/core/core.h...
Checking patch lib/cpluff/examples/cpfile/plugins/core/plugin.xml...
Checking patch lib/cpluff/examples/cpfile/plugins/extension/Makefile.am...
Checking patch lib/cpluff/examples/cpfile/plugins/extension/Makefile.nmake...
Checking patch lib/cpluff/examples/cpfile/plugins/extension/extension.c...
Checking patch lib/cpluff/examples/cpfile/plugins/extension/file_types.xsd...
Checking patch lib/cpluff/examples/cpfile/plugins/extension/plugin.xml...
Checking patch lib/cpluff/examples/cpfile/plugins/special/Makefile.am...
Checking patch lib/cpluff/examples/cpfile/plugins/special/Makefile.nmake...
Checking patch lib/cpluff/examples/cpfile/plugins/special/plugin.xml...
Checking patch lib/cpluff/examples/cpfile/plugins/special/special.c...
Checking patch lib/cpluff/examples/nmakedir.bat...
Checking patch lib/cpluff/kazlib/hash.c...
Checking patch lib/cpluff/kazlib/hash.h...
Checking patch lib/cpluff/kazlib/list.c...
Checking patch lib/cpluff/kazlib/list.h...
Checking patch lib/cpluff/libcpluff/Makefile.am...
Checking patch lib/cpluff/libcpluff/context.c...
Checking patch lib/cpluff/libcpluff/cpluff.c...
Checking patch lib/cpluff/libcpluff/cpluff.h...
Checking patch lib/cpluff/libcpluff/cpluffdef.h.in...
Checking patch lib/cpluff/libcpluff/defines.h...
Checking patch lib/cpluff/libcpluff/docsrc/Doxyfile-impl.in...
Checking patch lib/cpluff/libcpluff/docsrc/Doxyfile-ref.in...
Checking patch lib/cpluff/libcpluff/docsrc/Makefile.am...
Checking patch lib/cpluff/libcpluff/docsrc/architecture.dox...
Checking patch lib/cpluff/libcpluff/docsrc/mainpage.dox...
Checking patch lib/cpluff/libcpluff/docsrc/mainprog.dox...
Checking patch lib/cpluff/libcpluff/docsrc/plugin.dox...
Checking patch lib/cpluff/libcpluff/internal.h...
Checking patch lib/cpluff/libcpluff/logging.c...
Checking patch lib/cpluff/libcpluff/pcontrol.c...
Checking patch lib/cpluff/libcpluff/pinfo.c...
Checking patch lib/cpluff/libcpluff/ploader.c...
Checking patch lib/cpluff/libcpluff/pscan.c...
Checking patch lib/cpluff/libcpluff/psymbol.c...
Checking patch lib/cpluff/libcpluff/serial.c...
Checking patch lib/cpluff/libcpluff/thread.h...
Checking patch lib/cpluff/libcpluff/thread_posix.c...
Checking patch lib/cpluff/libcpluff/thread_windows.c...
Checking patch lib/cpluff/libcpluff/util.c...
Checking patch lib/cpluff/libcpluff/util.h...
Checking patch lib/cpluff/libcpluff/win32/cpluffdef.h...
Checking patch lib/cpluff/libcpluff/win32/dirent.c...
Checking patch lib/cpluff/libcpluff/win32/dirent.h...
Checking patch lib/cpluff/libcpluff/win32/win32_utils.c...
Checking patch lib/cpluff/libcpluff/win32/win32_utils.h...
Checking patch lib/cpluff/loader/Makefile.am...
Checking patch lib/cpluff/loader/loader.c...
Checking patch lib/cpluff/patches/0001-all-Change-plugin-to-addon-tags.patch...
Checking patch lib/cpluff/patches/0002-osx-Fix-include.patch...
Checking patch lib/cpluff/patches/0003-win32-Fix-compile.-Missing-cpluff.dll-will-crash-Kod.patch...
Checking patch lib/cpluff/patches/0004-all-Fix-log-level-inconsistencies.patch...
Checking patch lib/cpluff/patches/0005-all-Fix-unsigned-signed-warnings.patch...
Checking patch lib/cpluff/patches/0006-all-Fix-scoping-error.patch...
Checking patch lib/cpluff/patches/0007-all-Comment-log-line-to-reduce-log-spew.patch...
Checking patch lib/cpluff/patches/0008-configure-Handle-libdl-only-being-needed-on-Linux-fo.patch...
Checking patch lib/cpluff/patches/0009-configure-Add-silent-rules-to-shut-it-up-a-little.patch...
Checking patch lib/cpluff/patches/0010-configure-Fix-build-with-xcode5.patch...
Checking patch lib/cpluff/patches/0011-docs-We-don-t-need-them.patch...
Checking patch lib/cpluff/patches/0012-Kill-implicit-function-declaration-warning.patch...
Checking patch lib/cpluff/patches/0013-Kill-implicit-function-declaration-warning-2.patch...
Checking patch lib/cpluff/plugin.xsd...
Checking patch lib/cpluff/po/.gitignore...
Checking patch lib/cpluff/po/LINGUAS...
Checking patch lib/cpluff/po/Makevars...
Checking patch lib/cpluff/po/POTFILES.in...
Checking patch lib/cpluff/po/de.po...
Checking patch lib/cpluff/po/fi.po...
Checking patch lib/cpluff/test/Makefile.am...
Checking patch lib/cpluff/test/collections.c...
Checking patch lib/cpluff/test/cpinfo.c...
Checking patch lib/cpluff/test/expected/output-extensions.txt...
Checking patch lib/cpluff/test/expected/output-extpoints.txt...
Checking patch lib/cpluff/test/expected/output-loadmaximal.txt...
Checking patch lib/cpluff/test/expected/output-loadminimal.txt...
Checking patch lib/cpluff/test/extcfg.c...
Checking patch lib/cpluff/test/fatalerror.c...
Checking patch lib/cpluff/test/initdestroy.c...
Checking patch lib/cpluff/test/loggers.c...
Checking patch lib/cpluff/test/pcallbacks.c...
Checking patch lib/cpluff/test/pcollections/collection1/plugin1/plugin.xml...
Checking patch lib/cpluff/test/pcollections/collection1v2/plugin1/plugin.xml...
Checking patch lib/cpluff/test/pcollections/collection1v3/plugin1/plugin.xml...
Checking patch lib/cpluff/test/pcollections/collection2/plugin2a/plugin.xml...
Checking patch lib/cpluff/test/pcollections/collection2/plugin2b/plugin.xml...
Checking patch lib/cpluff/test/pcollections/dependencies/chain1/plugin.xml...
Checking patch lib/cpluff/test/pcollections/dependencies/chain2/plugin.xml...
Checking patch lib/cpluff/test/pcollections/dependencies/chain3/plugin.xml...
Checking patch lib/cpluff/test/pcollections/dependencies/chainmissingdep/plugin.xml...
Checking patch lib/cpluff/test/pcollections/dependencies/loop1/plugin.xml...
Checking patch lib/cpluff/test/pcollections/dependencies/loop2/plugin.xml...
Checking patch lib/cpluff/test/pcollections/dependencies/loop3/plugin.xml...
Checking patch lib/cpluff/test/pcollections/dependencies/loop4/plugin.xml...
Checking patch lib/cpluff/test/pcollections/dependencies/loop5/plugin.xml...
Checking patch lib/cpluff/test/pcollections/dependencies/missingdep/plugin.xml...
Checking patch lib/cpluff/test/pcollections/dependencies/sloop1/plugin.xml...
Checking patch lib/cpluff/test/pcollections/dependencies/sloop2/plugin.xml...
Checking patch lib/cpluff/test/pdependencies.c...
Checking patch lib/cpluff/test/pinstallation.c...
Checking patch lib/cpluff/test/ploading.c...
Checking patch lib/cpluff/test/plugins-source/Makefile.am...
Checking patch lib/cpluff/test/plugins-source/callbackcounter/Makefile.am...
Checking patch lib/cpluff/test/plugins-source/callbackcounter/callbackcounter.c...
Checking patch lib/cpluff/test/plugins-source/callbackcounter/callbackcounter.h...
Checking patch lib/cpluff/test/plugins-source/callbackcounter/plugin.xml...
Checking patch lib/cpluff/test/plugins-source/symprovider/Makefile.am...
Checking patch lib/cpluff/test/plugins-source/symprovider/plugin.xml...
Checking patch lib/cpluff/test/plugins-source/symprovider/symprovider.c...
Checking patch lib/cpluff/test/plugins-source/symuser/Makefile.am...
Checking patch lib/cpluff/test/plugins-source/symuser/plugin.xml...
Checking patch lib/cpluff/test/plugins-source/symuser/symuser.c...
Checking patch lib/cpluff/test/plugins/maximal/plugin.xml...
Checking patch lib/cpluff/test/plugins/minimal/plugin.xml...
Checking patch lib/cpluff/test/pscanning.c...
Checking patch lib/cpluff/test/psymbolusage.c...
Checking patch lib/cpluff/test/test-extensions...
Checking patch lib/cpluff/test/test-extpoints...
Checking patch lib/cpluff/test/test-fatalerrordefault...
Checking patch lib/cpluff/test/test-fatalerrorreset...
Checking patch lib/cpluff/test/test-loadmaximal...
Checking patch lib/cpluff/test/test-loadminimal...
Checking patch lib/cpluff/test/test.h...
Checking patch lib/cpluff/test/testmain.c...
Checking patch lib/cpluff/test/tests.txt...
Checking patch xbmc/addons/Addon.h...
Checking patch xbmc/addons/AddonBuilder.cpp...
Checking patch xbmc/addons/AddonBuilder.h...
Checking patch xbmc/addons/AddonDatabase.cpp...
Checking patch xbmc/addons/AddonDatabase.h...
Checking patch xbmc/addons/AddonManager.cpp...
Hunk #1 succeeded at 30 (offset -2 lines).
Hunk #2 succeeded at 37 (offset -2 lines).
Hunk #3 succeeded at 101 (offset -2 lines).
Hunk #4 succeeded at 128 (offset -2 lines).
error: while searching for:
  m_database.GetBlacklisted(tmp);
  m_updateBlacklist = std::move(tmp);

  bool result = false;
  if (m_cp_context)
  {
    result = true;
    cp_scan_plugins(m_cp_context, CP_SP_UPGRADE);
  }

  return result;
}

bool CAddonMgr::UnloadAddon(const std::string& addonId)

error: patch failed: xbmc/addons/AddonManager.cpp:673
error: xbmc/addons/AddonManager.cpp: patch does not apply
Checking patch xbmc/addons/AddonManager.h...
Checking patch xbmc/addons/ContextMenuAddon.cpp...
Checking patch xbmc/addons/ContextMenuAddon.h...
Checking patch xbmc/addons/FontResource.cpp...
Checking patch xbmc/addons/FontResource.h...
Checking patch xbmc/addons/GameResource.cpp...
Checking patch xbmc/addons/GameResource.h...
Checking patch xbmc/addons/ImageResource.cpp...
Checking patch xbmc/addons/ImageResource.h...
Checking patch xbmc/addons/LanguageResource.cpp...
Checking patch xbmc/addons/LanguageResource.h...
Checking patch xbmc/addons/PluginSource.cpp...
Checking patch xbmc/addons/PluginSource.h...
Checking patch xbmc/addons/Repository.cpp...
Checking patch xbmc/addons/Repository.h...
Checking patch xbmc/addons/Scraper.cpp...
Checking patch xbmc/addons/Scraper.h...
Checking patch xbmc/addons/Service.cpp...
Checking patch xbmc/addons/Service.h...
Checking patch xbmc/addons/Skin.cpp...
Checking patch xbmc/addons/Skin.h...
Checking patch xbmc/addons/Webinterface.cpp...
Checking patch xbmc/addons/Webinterface.h...
Checking patch xbmc/addons/addoninfo/AddonInfo.h...
Checking patch xbmc/addons/test/CMakeLists.txt...
Checking patch xbmc/addons/test/TestAddonBuilder.cpp...
Checking patch xbmc/addons/test/TestAddonFactory.cpp...
Checking patch xbmc/games/addons/GameClient.cpp...
Checking patch xbmc/games/addons/GameClient.h...
Checking patch xbmc/games/controllers/Controller.cpp...
Checking patch xbmc/games/controllers/Controller.h...
Checking patch xbmc/addons/addoninfo/AddonInfo.cpp...
Checking patch xbmc/addons/addoninfo/AddonInfo.h...
Checking patch xbmc/addons/addoninfo/AddonInfoBuilder.cpp...
Checking patch xbmc/addons/addoninfo/AddonInfoBuilder.h...
Checking patch xbmc/utils/Locale.cpp...
Checking patch xbmc/utils/Locale.h...
Checking patch xbmc/addons/test/CMakeLists.txt...
Checking patch xbmc/addons/test/TestAddonBuilder.cpp...
Checking patch xbmc/addons/test/TestAddonDatabase.cpp...
Checking patch xbmc/addons/test/TestAddonInfoBuilder.cpp...
```

AlwinEsch commented:
Is rebased.

MilhouseVH commented:
Thanks - applies now! I'll include this in nightly test builds and post any feedback.

MilhouseVH commented:
Apologies for the delay in testing this PR with nightlies - normal service should resume tonight! :smile:

ksooo commented on xbmc/addons/AddonManager.cpp:
```diff
@@ -815,6 +815,20 @@ void CAddonMgr::FindAddons(ADDON_INFO_LIST& addonmap, const std::string& path)
+                          __FUNCTION__, addonInfo->ID(), it->second->Path(), it->second->Version().asString(), addonInfo->Path(), addonInfo->Version().asString());
+              continue;
+            }
+            CLog::Log(LOGDEBUG, "CAddonMgr::{}: Addon '{}' already present on '{}' with version {} becomes ignored and replaced with version {} on '{}'",
```
I'm not a native English speaker, but for it feels right to just omit the word "becomes" from this string.

Rechi commented on xbmc/addons/addoninfo/AddonInfo.h:
```diff
@@ -8,173 +9,146 @@
+  DependencyInfo(std::string id, AddonVersion requiredVersion, bool optional)
+      : id(id), requiredVersion(requiredVersion), optional(optional) {}
+
+  bool operator==(const DependencyInfo &rhs) const
```
left align `&`

AlwinEsch replied:
Is changed here and on other places.

Rechi commented on xbmc/addons/addoninfo/AddonInfo.h:
```diff
@@ -8,173 +9,146 @@
+            optional == rhs.optional;
+  }
+
+  bool operator!=(const DependencyInfo &rhs) const
```
left align `&`

Rechi commented on xbmc/addons/addoninfo/AddonInfo.h:
```diff
@@ -8,173 +9,146 @@
+    * @brief Utilities to translate add-on parts to his requested part.
+    */
+  //@{
+  static std::string TranslateType(TYPE type, bool pretty=false);
```
space before and after `=`

Rechi commented on xbmc/addons/addoninfo/AddonInfo.h:
```diff
@@ -8,173 +9,146 @@
+  //@{
+  static std::string TranslateType(TYPE type, bool pretty=false);
+  static std::string TranslateIconType(TYPE type);
+  static TYPE TranslateType(const std::string &string);
```
left align `&`

Rechi commented on xbmc/addons/addoninfo/AddonInfo.h:
```diff
@@ -8,173 +9,146 @@
+  static std::string TranslateType(TYPE type, bool pretty=false);
+  static std::string TranslateIconType(TYPE type);
+  static TYPE TranslateType(const std::string &string);
+  static TYPE TranslateSubContent(const std::string &content);
```
left align `&`

Rechi commented on xbmc/addons/addoninfo/AddonInfoBuilder.cpp:
```diff
@@ -0,0 +1,668 @@
+      element = child->FirstChildElement("news");
+      while (element)
+      {
+        const char *lang = element->Attribute("lang");
```
left align `*`

Rechi commented on xbmc/addons/addoninfo/AddonInfo.h:
```diff
@@ -79,20 +80,20 @@ class CAddonInfo
   const ArtMap& Art() const { return m_art; }
   const std::vector<std::string>& Screenshots() const { return m_screenshots; }
-  const std::string& Disclaimer() const { return m_disclaimer; }
+  const std::string& Disclaimer() const { return  GetTranslatedText(m_disclaimer); }
```
single space is enough

Rechi commented on xbmc/addons/test/TestAddonInfoBuilder.cpp:
```diff
@@ -0,0 +1,168 @@
+using namespace ADDON;
+
+const std::string addonXML =
+"<addon id=\"metadata.blablabla.org\""
```
make this a raw string literal

MilhouseVH commented on xbmc/addons/AddonManager.cpp:
```diff
@@ -815,6 +815,20 @@ void CAddonMgr::FindAddons(ADDON_INFO_LIST& addonmap, const std::string& path)
+            }
+            CLog::Log(LOGDEBUG, "CAddonMgr::{}: Addon '{}' already present on '{}' with version {} becomes ignored and replaced with version {} on '{}'",
+                          __FUNCTION__, addonInfo->ID(), it->second->Path(), it->second->Version().asString(), addonInfo->Version().asString(), addonInfo->Path());
+          }
```
Not sure where @ksooo's comment went, but I'd suggest the following:
```
               CLog::Log(LOGWARNING, "CAddonMgr::{}: Addon '{}' already present with higher version {} at '{}' - other version {} at '{}' will be ignored",
                           __FUNCTION__, addonInfo->ID(), it->second->Version().asString(), , it->second->Path(), addonInfo->Version().asString(), addonInfo->Path());
```
and
```
             CLog::Log(LOGDEBUG, "CAddonMgr::{}: Addon '{}' already present with version {} at '{}' will be replaced with version {} at '{}'",
                           __FUNCTION__, addonInfo->ID(), it->second->Version().asString(), it->second->Path(), addonInfo->Version().asString(), addonInfo->Path());
```

This consistently logs the version followed by path each time (I think I got the arguments correct!).

ksooo replied:
> Not sure where @ksooo's comment went

I accidentally deleted my comment. Thanks for following up.

AlwinEsch replied:
Thanks is changed!

MilhouseVH commented on xbmc/addons/AddonManager.cpp:
```diff
@@ -815,6 +815,20 @@ void CAddonMgr::FindAddons(ADDON_INFO_LIST& addonmap, const std::string& path)
+          {
+            if (it->second->Version() > addonInfo->Version())
+            {
+
```
I guess remove this line ^^

AlwinEsch committed:
[addons] add check if same addon on two places

This check the addons if they are present on two different
places and takes only the highest version.

MilhouseVH commented:
Can this be rebased (rotted by #16247)

AlwinEsch commented:
Rebase is done :smile:, how is the state to bring in?

MilhouseVH commented:
> how is the state to bring in?

No issues reported so far! LGTM.

Rechi commented:
CI reports errors, that has to be fixed first.

AlwinEsch commented:
Where can I find the CI reports errors? 
Also do I not find the exact place where Jenkins fail on Mac and Android, the builds was OK but seems to make a warning to error, but which?

lrusak commented:
> Where can I find the CI reports errors?
> Also do I not find the exact place where Jenkins fail on Mac and Android, the builds was OK but seems to make a warning to error, but which?

https://jenkins.kodi.tv/job/OSX-64/14285/warnings2Result/new/

AlwinEsch committed:
[addons] adds support for various description languages

The fact that the list of addons is stored in the manager, it is not easy
to get the appropriate language. As the language is not yet known when
reading the Replace list and the list would not be correct when changing
the language.

As in the past, the database will be used there, but unfortunately, after
a language change, the database text will continue to be incorrect until
it has been updated again.

AlwinEsch commented:
Thanks @lrusak, thought is somewhere on a TravisCI build about a memleak report or somethink else and has not found a mail or Link.

AlwinEsch commented:
@Rechi build has worked now

Rechi commented:
The following lines can be removed, as that expat version is only used by cpluff.
https://github.com/xbmc/xbmc/blob/cba67c6302352e5c4347f0454b51166908ec6691/project/BuildDependencies/scripts/0_package.target-win10-arm.list#L11
https://github.com/xbmc/xbmc/blob/cba67c6302352e5c4347f0454b51166908ec6691/project/BuildDependencies/scripts/0_package.target-win10-win32.list#L11
https://github.com/xbmc/xbmc/blob/cba67c6302352e5c4347f0454b51166908ec6691/project/BuildDependencies/scripts/0_package.target-win10-x64.list#L11
https://github.com/xbmc/xbmc/blob/15c04551a7746b2495ea4510fd16902147432658/project/BuildDependencies/scripts/0_package.target-win32.list#L13
https://github.com/xbmc/xbmc/blob/b21f6671dac47222361a10faa82336a7ff5db70b/project/BuildDependencies/scripts/0_package.target-x64.list#L13

AlwinEsch commented:
@Rechi have started to remove the expat, but is it not needed by Python?

see https://github.com/xbmc/xbmc/blob/master/cmake/modules/FindPython.cmake#L18

Rechi commented:
Are the Windows (== !KODI_DEPENDSBUILD) failing?

AlwinEsch commented:
Have checked, build and run has worked on Windows.

AlwinEsch committed:
[depends] remove expat from win build

MilhouseVH commented:
This has been in test builds for a while, but is now broken... is this merging any time soon? I don't intend testing it any longer so no rush on fixing it from my POV. :)

AlwinEsch committed:
[addons] remove cpluff complete

AlwinEsch commented:
Is rebased and tested.

How is now the state to bring in? From my side OK, but I not merge this big change without a approval from another :cold_sweat: (by e.g. "OK to merge in now" :relieved:).

Rechi commented:
jenkins build this please

lrusak commented:
Do it 👍

Rechi commented on xbmc/addons/binary-addons/BinaryAddonBase.h:
```diff
@@ -10,7 +10,8 @@
 #include "AddonInstanceHandler.h"
 #include "BinaryAddonType.h"
-#include "addons/AddonInfo.h"
+
```
remove the empty line

Rechi commented on xbmc/addons/binary-addons/BinaryAddonType.h:
```diff
@@ -9,7 +9,8 @@
 
 #include "BinaryAddonExtensions.h"
-#include "addons/AddonInfo.h"
+
```
remove the empty line

Rechi commented on xbmc/addons/addoninfo/AddonInfo.h:
```diff
@@ -1,3 +1,4 @@
+
```
remove the empty line

Rechi commented on xbmc/addons/addoninfo/AddonInfo.h:
```diff
@@ -8,172 +9,146 @@
 #pragma once
 
+#include "AddonType.h"
+
```
remove the empty line

Rechi commented on xbmc/addons/addoninfo/AddonInfoBuilder.cpp:
```diff
@@ -0,0 +1,668 @@
+
+#include "AddonInfoBuilder.h"
+#include "AddonType.h"
+
```
move the empty line one line above

Rechi commented on xbmc/addons/addoninfo/AddonInfoBuilder.cpp:
```diff
@@ -0,0 +1,668 @@
+#include "LangInfo.h"
+#include "filesystem/File.h"
+#include "filesystem/SpecialProtocol.h"
+#include "utils/log.h"
```
move this line at the end of this block

Rechi commented on xbmc/addons/addoninfo/AddonType.cpp:
```diff
@@ -0,0 +1,34 @@
+
+#include "AddonType.h"
+#include "AddonInfo.h"
+
```
move the empty line one line above

Rechi commented on xbmc/addons/test/TestAddonInfoBuilder.cpp:
```diff
@@ -0,0 +1,169 @@
+
+#include "addons/addoninfo/AddonInfoBuilder.h"
+
+#include "gtest/gtest.h"
```
wrong include style and order

AlwinEsch committed:
[addons] add primitives to create CAddonInfo without cpluff

This adds the necessary elements to create CAddonInfo without cpluff. This code is currently not used and comes in the following changes.

AlwinEsch committed:
[addons] move AddonInfo (.cpp / .h) to its own folder

This is done to make it easier to understand after the subsequent changes and to have all of this related files in a folder.

AlwinEsch committed:
[addons] changing the test system in a new way without cpluff

This rewrites existing test things in the new way and adds a test to create the CAddonInfo class.

AlwinEsch commented:
jenkins build this please

@Rechi have changed your comment places, now OK for you?

AlwinEsch commented:
I do the merge tomorrow morning, if nothing else against it

Rechi commented:
@AlwinEsch please always set the milestone before or right after merging a PR.

Rechi commented:
This broke installing (and upadeting) binary add-ons from the official Kodi repo on all supported platforms (android, osx and desktop widows).

see https://forum.kodi.tv/showthread.php?tid=345851 and https://forum.kodi.tv/showthread.php?tid=345994

AlwinEsch commented:
I'm checking right now, unfortunately, in the forum is not directly apparent from which OS it comes, further no kodi.log were given which makes it hard to recognize the cause.

On libreelec unfortunately the https://addons.libreelec.tv use failed, both on Kodi 19.0-ALPHA1, but also with Kodi 18.3.

Now I look at Windows what is coming.

CvH commented:
> On libreelec unfortunately the https://addons.libreelec.tv use failed, both on Kodi 19.0-ALPHA1, but also with Kodi 18.3.

what failed ?

tmm1 commented on xbmc/addons/AddonManager.cpp:
```diff
@@ -984,242 +671,12 @@ bool CAddonMgr::IsSystemAddon(const std::string& id)
-{
-  std::string libraryName;
-#if defined(TARGET_ANDROID)
-  libraryName = GetExtValue(base, "@library_android");
```
Previously an `addon.xml` could include entries like `library_android=`, `library_windows` to point to different binary addon dlls for different platforms all in the same zip file. It looks like this functionality was removed completely? Was that intentional?

AlwinEsch replied:
Is still possible, this now only on other place now https://github.com/xbmc/xbmc/blob/master/xbmc/addons/addoninfo/AddonInfoBuilder.cpp#L537-L566

And addon.xml can look like, them:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<addon
  id="visualization.shadertoy"
  version="2.0.0"
  name="Shadertoy"
  provider-name="Team Kodi">
  <requires>@ADDON_DEPENDS@</requires>
  <extension
    point="xbmc.player.musicviz"
    library_windows="visualization.shadertoy-win64.dll"
    library_osx="visualization.shadertoy-mac64.dylib"
    library_linux=""visualization.shadertoy-linux64.so"/>"/>
  <extension point="xbmc.addon.metadata">
    <summary lang="en">Visualizations from Shadertoy</summary>
    <description lang="en">This music visualization is based on Shadertoy based GPU programs.</description>
    <platform>windows-x86_64 osx-x86_64 linux</platform> <!-- Here to set manually the supported platforms -->
  </extension>
</addon>
```

About platform visible here https://github.com/xbmc/xbmc/blob/master/xbmc/addons/addoninfo/AddonInfoBuilder.cpp#L568-L628

Only was it a very long time not used with binary addon where have different libs together, at that time Kodi supported only 32bit x86.

Thats why no cpu in `GetPlatformLibraryName` and on xml by `platform` only linux, but also because Linux might not work 100% in other places (for example, built under Ubuntu, used on Fedora), not as easy as Mac or Windows.

tmm1 replied:
Aha, thanks for the link. My search failed me because github had collapsed some of the files in the diff.
